File: apps/android/app/src/main/java/com/galaxyssi/chat/MqttDeliveryEnvelope.kt
package com.galaxyssi.chat

import org.json.JSONObject
import java.security.MessageDigest

/** Metadata lives inside existing pair AEAD. Parsing alone never authenticates or stores a message. */
internal object MqttDeliveryEnvelope {
    const val FIELD = "_mqtt_delivery"
    const val ALGORITHM = "signal-wire-sha256-v1"
    const val RECEIPT_TYPE = "link_rx_stored"
    const val MAX_SAFE_INTEGER = 9_007_199_254_740_991L
    private const val MAX_WIRE_BODY = 4 * 1024 * 1024
    private val hash = Regex("[a-f0-9]{64}")
    private val token = Regex("[a-f0-9]{32}")
    private val traffic = setOf("control", "message", "final", "progress", "chunk", "receipt")
    private val textFields = listOf("scheme", "from", "to", "signal_type", "type", "body", "protocol")
    private val numberFields = listOf("message_type", "messageType", "device_id", "version")

    data class Message(val messageId: String, val contentHash: String, val sender: String,
                       val receiver: String, val traffic: String) {
        init {
            checkedString(messageId)
            require(hash.matches(contentHash) && hash.matches(sender) && hash.matches(receiver))
            require(sender != receiver && traffic in MqttDeliveryEnvelope.traffic)
        }
    }

    data class Attempt(val attemptId: String, val brokerId: String, val generation: Long) {
        init {
            require(token.matches(attemptId) && brokerId in MqttBrokerCatalog.brokers)
            require(generation in 1..MAX_SAFE_INTEGER)
        }
    }

    data class Frame(val message: Message, val attempt: Attempt) {
        fun metadata(): JSONObject = JSONObject().put("version", 1).put("content_hash_algorithm", ALGORITHM)
            .put("message_id", message.messageId).put("content_hash", message.contentHash)
            .put("sender", message.sender).put("receiver", message.receiver).put("traffic", message.traffic)
            .put("attempt_id", attempt.attemptId).put("broker_id", attempt.brokerId).put("generation", attempt.generation)

        fun attach(wire: JSONObject): JSONObject {
            require(!wire.has(FIELD) && contentHash(wire) == message.contentHash)
            return JSONObject(wire.toString()).put(FIELD, metadata())
        }

        fun receiptAfterStore(storedMessageId: String, storedContentHash: String): JSONObject {
            validateApplication(storedMessageId, storedContentHash)
            require(message.traffic != "receipt") { "Receipts cannot request receipts" }
            return metadata().put("type", RECEIPT_TYPE).put("status", "RX_STORED")
        }

        fun validateApplication(messageId: String, wireHash: String) {
            require(message.messageId == messageId && message.contentHash == wireHash) {
                "Delivery frame does not match Signal application message"
            }
        }
    }

    /** UTF-8 length-prefixed Signal fields, independent of JSON order, numbers, or UTF-16 length. */
    fun contentHash(wire: JSONObject): String {
        require(wire.opt("scheme") == "signal")
        for (key in listOf("from", "to", "body")) {
            checkedString(wire.opt(key), if (key == "body") MAX_WIRE_BODY else 512)
        }
        val digest = MessageDigest.getInstance("SHA-256")
        digest.update("GalaxySSI/SignalWireReceipt/v1\u0000".toByteArray(Charsets.UTF_8))
        for (key in textFields + numberFields) {
            if (!wire.has(key)) continue
            val value = wire.get(key)
            val kind: String
            val encoded: ByteArray
            if (key in textFields) {
                kind = "s"
                encoded = checkedString(value, if (key == "body") MAX_WIRE_BODY else 512).toByteArray(Charsets.UTF_8)
            } else {
                kind = "i"
                encoded = checkedInteger(value).toString().toByteArray(Charsets.US_ASCII)
            }
            digest.update("${key.length}:$key$kind${encoded.size}:".toByteArray(Charsets.US_ASCII))
            digest.update(encoded)
        }
        return digest.digest().joinToString("") { "%02x".format(it) }
    }

    fun parseVerifiedFrame(wire: JSONObject, sender: String, receiver: String, ingressBroker: String): Frame {
        val frame = parseMetadata(wire.getJSONObject(FIELD))
        require(frame.message.sender == sender && frame.message.receiver == receiver && frame.attempt.brokerId == ingressBroker)
        require(frame.message.contentHash == contentHash(wire)) { "Delivery ciphertext digest mismatch" }
        return frame
    }

    /** The authenticated actor is originalReceiver; the ACK may return on another path. */
    fun parseVerifiedReceipt(payload: JSONObject, originalSender: String, originalReceiver: String): Frame {
        require(payload.opt("type") == RECEIPT_TYPE && payload.opt("status") == "RX_STORED")
        return parseMetadata(payload).also {
            require(it.message.sender == originalSender && it.message.receiver == originalReceiver)
            require(it.message.traffic != "receipt")
        }
    }

    fun receiptBinding(scope: String, sender: String, receiver: String, secret: String): String {
        val digest = MessageDigest.getInstance("SHA-256")
        digest.update("GalaxySSI/OutboundReceiptBinding/v1\u0000".toByteArray(Charsets.UTF_8))
        for (value in listOf(scope, sender, receiver, secret)) {
            val encoded = checkedString(value, 512).toByteArray(Charsets.UTF_8)
            digest.update("${encoded.size}:".toByteArray(Charsets.US_ASCII))
            digest.update(encoded)
        }
        return digest.digest().joinToString("") { "%02x".format(it) }
    }

    fun storedReceipt(messageId: String, wireHash: String): JSONObject {
        checkedString(messageId)
        require(hash.matches(wireHash))
        return JSONObject().put("type", "delivery_ack").put("delivery_status", "RX_STORED")
            .put("transport_message_id", messageId).put("content_hash", wireHash).put("content_hash_algorithm", ALGORITHM)
    }

    fun parseStoredReceipt(payload: JSONObject): Pair<String, String> {
        require(payload.opt("type") == "delivery_ack" && payload.opt("delivery_status") == "RX_STORED" &&
            payload.opt("content_hash_algorithm") == ALGORITHM)
        val message = checkedString(payload.opt("transport_message_id"))
        val digest = checkedString(payload.opt("content_hash"))
        require(hash.matches(digest))
        return message to digest
    }

    private fun parseMetadata(value: JSONObject): Frame {
        require(checkedInteger(value.opt("version")) == 1L && value.opt("content_hash_algorithm") == ALGORITHM)
        return Frame(Message(checkedString(value.opt("message_id")), checkedString(value.opt("content_hash")),
            checkedString(value.opt("sender")), checkedString(value.opt("receiver")), checkedString(value.opt("traffic"))),
            Attempt(checkedString(value.opt("attempt_id")), checkedString(value.opt("broker_id")),
                checkedInteger(value.opt("generation"))))
    }

    private fun checkedString(value: Any?, maximum: Int = 256): String {
        require(value is String && value.isNotEmpty() && value.toByteArray(Charsets.UTF_8).size <= maximum)
        require(value.none { it.code < 32 || it.code == 127 })
        require(value.indices.none { index ->
            val char = value[index]
            (char.isHighSurrogate() && (index == value.lastIndex || !value[index + 1].isLowSurrogate())) ||
                (char.isLowSurrogate() && (index == 0 || !value[index - 1].isHighSurrogate()))
        })
        return value
    }

    private fun checkedInteger(value: Any?): Long {
        require(value is Byte || value is Short || value is Int || value is Long)
        return (value as Number).toLong().also { require(it in 1..MAX_SAFE_INTEGER) }
    }
}

File: apps/desktop/core/galaxyssi-link/backend/mqtt_delivery_envelope.py
"""Authenticated attempt metadata around immutable Signal ciphertext.

Use only inside Link's existing authenticated outer envelope. This module does
not authenticate a packet or authorize a route. A receipt may be constructed
only from a committed inbox record, never from a successful MQTT publication.
"""
from __future__ import annotations

from dataclasses import dataclass
import hashlib
import re

from mqtt_broker_catalog import BROKER_IDS

FIELD = "_mqtt_delivery"
VERSION = 1
ALGORITHM = "signal-wire-sha256-v1"
RECEIPT_TYPE = "link_rx_stored"
MAX_WIRE_BODY = 4 * 1024 * 1024
MAX_SAFE_INTEGER = 9_007_199_254_740_991
HASH = re.compile(r"[a-f0-9]{64}")
TOKEN = re.compile(r"[a-f0-9]{32}")
TRAFFIC = frozenset({"control", "message", "final", "progress", "chunk", "receipt"})
TEXT_FIELDS = ("scheme", "from", "to", "signal_type", "type", "body", "protocol")
NUMBER_FIELDS = ("message_type", "messageType", "device_id", "version")


def _string(value, *, maximum=256):
    if (not isinstance(value, str) or not value or len(value.encode("utf-8")) > maximum
            or any(ord(char) < 32 or ord(char) == 127 for char in value)):
        raise ValueError("Invalid delivery string")
    return value


def _hash(value):
    if not isinstance(value, str) or not HASH.fullmatch(value):
        raise ValueError("Invalid delivery digest")
    return value


def _integer(value):
    if type(value) is not int or not 0 < value <= MAX_SAFE_INTEGER:
        raise ValueError("Invalid delivery integer")
    return value


def content_hash(wire: dict) -> str:
    """Cross-runtime digest of Signal fields, not reserialized application JSON.

    Domain-separated UTF-8 length prefixes avoid language-dependent JSON key
    order, floating-point formatting, and UTF-16 character-count differences.
    Mutable wall-clock/transport fields are intentionally excluded. A changed
    Signal ciphertext (including a session repair) gets a different digest.
    """
    if not isinstance(wire, dict) or wire.get("scheme") != "signal":
        raise ValueError("A Signal wire envelope is required")
    for key in ("from", "to", "body"):
        _string(wire.get(key), maximum=MAX_WIRE_BODY if key == "body" else 512)
    digest = hashlib.sha256(b"GalaxySSI/SignalWireReceipt/v1\0")
    for key in (*TEXT_FIELDS, *NUMBER_FIELDS):
        if key not in wire:
            continue
        value = wire[key]
        if key in TEXT_FIELDS:
            encoded = _string(value, maximum=MAX_WIRE_BODY if key == "body" else 512).encode("utf-8")
            kind = b"s"
        else:
            encoded = str(_integer(value)).encode("ascii")
            kind = b"i"
        digest.update(f"{len(key)}:{key}".encode("ascii") + kind + str(len(encoded)).encode("ascii") + b":" + encoded)
    return digest.hexdigest()


@dataclass(frozen=True)
class Message:
    message_id: str
    content_hash: str
    sender: str
    receiver: str
    traffic: str

    def __post_init__(self):
        _string(self.message_id)
        _hash(self.content_hash)
        _hash(self.sender)
        _hash(self.receiver)
        if self.sender == self.receiver or not isinstance(self.traffic, str) or self.traffic not in TRAFFIC:
            raise ValueError("Invalid delivery message identity")


@dataclass(frozen=True)
class Attempt:
    attempt_id: str
    broker_id: str
    generation: int

    def __post_init__(self):
        if not isinstance(self.attempt_id, str) or not TOKEN.fullmatch(self.attempt_id):
            raise ValueError("Invalid delivery attempt")
        if not isinstance(self.broker_id, str) or self.broker_id not in BROKER_IDS:
            raise ValueError("Unknown delivery broker")
        _integer(self.generation)


@dataclass(frozen=True)
class Frame:
    message: Message
    attempt: Attempt

    def metadata(self):
        return {"version": VERSION, "content_hash_algorithm": ALGORITHM,
                "message_id": self.message.message_id, "content_hash": self.message.content_hash,
                "sender": self.message.sender, "receiver": self.message.receiver,
                "traffic": self.message.traffic, "attempt_id": self.attempt.attempt_id,
                "broker_id": self.attempt.broker_id, "generation": self.attempt.generation}

    def attach(self, wire):
        if FIELD in wire or content_hash(wire) != self.message.content_hash:
            raise ValueError("Attempt does not match immutable ciphertext")
        return {**wire, FIELD: self.metadata()}

    def receipt_after_store(self, *, stored_message_id, stored_content_hash):
        if (stored_message_id, stored_content_hash) != (self.message.message_id, self.message.content_hash):
            raise ValueError("Attempt is not bound to the committed inbox record")
        if self.message.traffic == "receipt":
            raise ValueError("Receipts cannot request further receipts")
        return {"type": RECEIPT_TYPE, "status": "RX_STORED", **self.metadata()}


def _parse_metadata(value):
    if not isinstance(value, dict) or type(value.get("version")) is not int or value["version"] != VERSION:
        raise ValueError("Invalid delivery metadata version")
    if value.get("content_hash_algorithm") != ALGORITHM:
        raise ValueError("Unsupported delivery digest algorithm")
    return Frame(Message(value.get("message_id"), value.get("content_hash"), value.get("sender"),
                         value.get("receiver"), value.get("traffic")),
                 Attempt(value.get("attempt_id"), value.get("broker_id"), value.get("generation")))


def parse_verified_frame(wire, *, sender, receiver, ingress_broker):
    """Call after pair AEAD authentication; no dedup writes occur on failure."""
    if not isinstance(wire, dict) or FIELD not in wire:
        raise ValueError("Missing delivery attempt")
    frame = _parse_metadata(wire[FIELD])
    if (frame.message.sender, frame.message.receiver, frame.attempt.broker_id) != (sender, receiver, ingress_broker):
        raise ValueError("Delivery attempt belongs to another pair or path")
    if frame.message.content_hash != content_hash(wire):
        raise ValueError("Delivery ciphertext digest mismatch")
    return frame


def parse_verified_receipt(payload, *, original_sender, original_receiver):
    """Call after authenticating the receipt sender as original_receiver.

    Sender generation describes the original publication, not the ACK ingress
    generation. An ACK is allowed to return over another authenticated path.
    The dispatcher must still match a locally reserved attempt before acting.
    """
    if not isinstance(payload, dict) or payload.get("type") != RECEIPT_TYPE or payload.get("status") != "RX_STORED":
        raise ValueError("Not a durable receive receipt")
    frame = _parse_metadata(payload)
    if (frame.message.sender, frame.message.receiver) != (original_sender, original_receiver):
        raise ValueError("Receipt belongs to another pair")
    if frame.message.traffic == "receipt":
        raise ValueError("Receipt loops are forbidden")
    return frame


def receipt_binding(scope, sender, receiver, secret):
    digest = hashlib.sha256(b"GalaxySSI/OutboundReceiptBinding/v1\0")
    for value in (scope, sender, receiver, secret):
        encoded = _string(value, maximum=512).encode("utf-8")
        digest.update(str(len(encoded)).encode("ascii") + b":" + encoded)
    return digest.hexdigest()


def stored_receipt(message_id, wire_hash):
    return {"type": "delivery_ack", "delivery_status": "RX_STORED",
            "transport_message_id": _string(message_id), "content_hash": _hash(wire_hash),
            "content_hash_algorithm": ALGORITHM}


def parse_stored_receipt(payload):
    if (not isinstance(payload, dict) or payload.get("type") != "delivery_ack"
            or payload.get("delivery_status") != "RX_STORED" or payload.get("content_hash_algorithm") != ALGORITHM):
        raise ValueError("Unverified durable receipt semantics")
    return _string(payload.get("transport_message_id")), _hash(payload.get("content_hash"))
